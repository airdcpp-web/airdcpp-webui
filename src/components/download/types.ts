import * as UI from 'types/ui';


export type BrowseHandler = (() => void) | undefined;

export type PathDownloadHandler = (targetPath: string, targetFilename?: string) => Promise<any>;

export type LayoutProps = React.PropsWithChildren<{
  menuItems: React.ReactNode[];
  title: string;
  handleBrowse: BrowseHandler;
  t: UI.TranslateF;
}>;
