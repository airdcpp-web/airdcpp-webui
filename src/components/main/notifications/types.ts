export type NotificationLevel = 'error' | 'warning' | 'info' | 'success';
