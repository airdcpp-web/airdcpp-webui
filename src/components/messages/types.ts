import { IntersectionObserverProps } from 'react-intersection-observer';

import * as UI from '@/types/ui';

export interface CommonMessageProps extends Omit<IntersectionObserverProps, 'children'> {
  highlightMenuProps: UI.MessageActionMenuData;
}
