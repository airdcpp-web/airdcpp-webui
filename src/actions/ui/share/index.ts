export * from './ShareActions';
export * from './ShareExcludeActions';
export * from './ShareProfileActions';
export * from './ShareRootActions';
