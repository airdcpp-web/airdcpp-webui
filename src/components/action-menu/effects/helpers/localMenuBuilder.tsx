import { toActionI18nKey } from '@/utils/ActionUtils';

import * as UI from '@/types/ui';

import { ActionClickHandler } from '@/decorators/ActionHandlerDecorator';
import { Trans } from 'react-i18next';
import { parseTranslationModules } from '@/utils/TranslationUtils';

const getDivider = (menuIndex: number, itemIndex: number) => {
  return {
    id: `divider${menuIndex}_${itemIndex}`,
  };
};

const actionToActionMenuItem = <
  ItemDataT extends UI.ActionMenuItemDataValueType,
  EntityT extends UI.ActionMenuItemEntityValueType,
>(
  action: UI.MenuActionDefition<ItemDataT, EntityT>,
  menu: UI.ActionMenuType<ItemDataT, EntityT>,
  onClickAction: ActionClickHandler<ItemDataT, EntityT>,
) => {
  const active = !action.checked ? false : action.checked(menu.itemDataGetter());
  const icon = !!action.checked ? (active ? 'checkmark' : '') : action.icon;
  return {
    id: action.id,
    item: {
      onClick: () => {
        onClickAction({
          action,
          itemData: menu.itemDataGetter(),
          entity: menu.entity,
          moduleData: menu.moduleData,
        });
      },
      active,
      icon,
      children: (
        <Trans
          i18nKey={toActionI18nKey(
            action,
            parseTranslationModules(menu.moduleData.moduleId),
          )}
          defaults={action.displayName}
        >
          {action.displayName}
        </Trans>
      ),
    },
  };
};

// Convert ID to menu link element
const getMenuItem = <
  ItemDataT extends UI.ActionMenuItemDataValueType,
  EntityT extends UI.ActionMenuItemEntityValueType,
>(
  menu: UI.ActionMenuType<ItemDataT, EntityT>,
  menuIndex: number,
  action: UI.MenuActionListItemType<ItemDataT, EntityT>,
  itemIndex: number,
  onClickAction: ActionClickHandler<ItemDataT, EntityT>,
): UI.ActionMenuItem => {
  if (!action) {
    return getDivider(menuIndex, itemIndex);
  }

  if ('children' in action) {
    const children = action.children.map((childAction, childActionIndex) => {
      if (!childAction) {
        return getDivider(0, childActionIndex);
      }

      return actionToActionMenuItem(childAction, menu, onClickAction);
    });

    if (children.length === 1) {
      return children[0];
    }

    return {
      id: action.id,
      item: {
        icon: action.icon,
        onClick: () => {
          // ..
        },
        children: <>{action.displayName}</>,
      },
      children,
    };
  }

  return actionToActionMenuItem(action, menu, onClickAction);
};

export const localMenuToActionMenuItems = <
  ItemDataT extends UI.ActionMenuItemDataValueType,
  EntityT extends UI.ActionMenuItemEntityValueType,
>(
  onClickAction: ActionClickHandler<ItemDataT, EntityT>,
  menu: UI.ActionMenuType<ItemDataT, EntityT>,
  menuIndex: number,
) => {
  const items = menu.actions.map((actionId, actionIndex) => {
    return getMenuItem(
      menu as UI.ActionMenuType<ItemDataT, EntityT>,
      menuIndex,
      actionId,
      actionIndex,
      onClickAction,
    );
  });

  return items;
};
