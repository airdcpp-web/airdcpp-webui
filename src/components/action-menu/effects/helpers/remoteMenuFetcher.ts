import { useEffect, useState } from 'react';

import * as API from 'types/api';
import * as UI from 'types/ui';

import MenuConstants, { MENU_SUPPORTS } from 'constants/MenuConstants';
import NotificationActions from 'actions/NotificationActions';
import { parseActionMenuItemId, parseActionMenuItemIds } from 'utils/MenuUtils';
import { ActionMenuDefinition } from '../useActionMenuItems';
import { useSocket } from 'context/SocketContext';

export interface RemoteMenuData {
  selectedIds: UI.ActionIdType[];
  remoteMenuId: string;
  entityId: API.IdType | undefined;
}

export interface RemoteMenu {
  menuData: RemoteMenuData;
  items: API.GroupedContextMenuItem[];
}

interface ParsedDefinitionArray {
  remoteMenuIds: Array<string | undefined>;
  selectedIdsByMenu: Array<UI.ActionIdType>[];
  entityIds: Array<API.IdType | undefined>;
}

const parseDefinitionArray = (
  definitionArray: ActionMenuDefinition<any, any>[],
): ParsedDefinitionArray => {
  const selectedIdsByMenu = definitionArray
    .map((defition) => defition.itemData)
    .map((data) => parseActionMenuItemIds(data));
  const remoteMenuIds = definitionArray.map(
    (remoteMenuProps) => remoteMenuProps.remoteMenuId,
  );
  const entityIds = definitionArray
    .map((defition) => defition.entity)
    .map((data) => parseActionMenuItemId(data) as API.IdType | undefined);

  return { selectedIdsByMenu, remoteMenuIds, entityIds };
};

export const useRemoteMenuFetcher = (
  definitionArray: ActionMenuDefinition<any, any>[],
) => {
  const { selectedIdsByMenu, remoteMenuIds, entityIds } =
    parseDefinitionArray(definitionArray);

  const [menusByType, setMenusByType] = useState<
    Array<RemoteMenu | undefined> | undefined
  >(undefined);

  const socket = useSocket();

  useEffect(() => {
    const fetchMenus = async () => {
      const menuPromises = remoteMenuIds.map(async (remoteMenuId, menuIndex) => {
        if (!remoteMenuId) {
          return Promise.resolve(undefined);
        }

        const selectedIds = selectedIdsByMenu[menuIndex];
        const entityId = entityIds[menuIndex];

        const menu = await socket.post<API.GroupedContextMenuItem[]>(
          `${MenuConstants.MODULE_URL}/${remoteMenuId}/list_grouped`,
          {
            selected_ids: selectedIds,
            entity_id: entityId,
            supports: MENU_SUPPORTS,
          },
        );

        return {
          items: menu,
          menuData: {
            remoteMenuId,
            selectedIds,
            entityId,
          },
        };
      });

      let fetchedMenus: Array<RemoteMenu | undefined>;
      try {
        fetchedMenus = await Promise.all(menuPromises);
      } catch (e) {
        NotificationActions.apiError('Failed fetch menu items', e);
        return;
      }

      setMenusByType(fetchedMenus);
    };

    fetchMenus();
  }, []);

  return menusByType;
};
