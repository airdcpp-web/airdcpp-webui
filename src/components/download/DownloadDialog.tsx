//import PropTypes from 'prop-types';
// import { useRef } from 'react';

import * as React from 'react';
import { Route, Routes, useNavigate } from 'react-router-dom';
import { useTranslation } from 'react-i18next';

import ShareConstants from 'constants/ShareConstants';
import {
  default as HistoryConstants,
  HistoryStringEnum,
} from 'constants/HistoryConstants';
import FavoriteDirectoryConstants from 'constants/FavoriteDirectoryConstants';
import FilesystemConstants from 'constants/FilesystemConstants';
import IconConstants from 'constants/IconConstants';

import NotificationActions from 'actions/NotificationActions';
import LoginStore from 'stores/LoginStore';

import DataProviderDecorator, {
  DataProviderDecoratorChildProps,
} from 'decorators/DataProviderDecorator';
import ModalRouteDecorator, {
  ModalRouteDecoratorChildProps,
} from 'decorators/ModalRouteDecorator';

import Modal, { ModalProps } from 'components/semantic/Modal';
import { FileBrowserDialog } from 'components/filebrowser';

import { runBackgroundSocketAction } from 'utils/ActionUtils';
import { toI18nKey, translate } from 'utils/TranslationUtils';
import { addHistory } from 'services/api/HistoryApi';

import { DownloadLayout, DownloadDataProps } from './layout';

import * as API from 'types/api';
import * as UI from 'types/ui';
import { PathDownloadHandler } from './types';

import './style.css';

export type DownloadDialogProps<
  ItemT extends UI.DownloadableItemInfo = UI.DownloadableItemInfo,
> = UI.ItemDownloadHandler<ItemT, Props<ItemT>>;

/*type DownloadItemIdType = string;

interface RouteProps {
  downloadItemId: DownloadItemIdType;
}*/

type DownloadDialogRouteProps = ModalRouteDecoratorChildProps;

interface DownloadDialogDataProps<
  ItemT extends UI.DownloadableItemInfo = UI.DownloadableItemInfo,
> extends DataProviderDecoratorChildProps,
    DownloadDataProps<ItemT> {}

type Props<ItemT extends UI.DownloadableItemInfo = UI.DownloadableItemInfo> =
  DownloadDialogProps<ItemT> & DownloadDialogDataProps<ItemT> & DownloadDialogRouteProps;

const DownloadDialog: React.FC<Props> = (props) => {
  const { t } = useTranslation();
  const navigate = useNavigate();

  const {
    downloadHandler,
    itemInfo,
    userGetter,
    params,
    session,
    historyPaths,
    favoritePaths,
    sharePaths,
    handleClose,
    ...other
  } = props;

  const handleDownload: PathDownloadHandler = async (targetPath, targetFilename) => {
    try {
      await downloadHandler(
        itemInfo,
        !!userGetter ? userGetter(params.downloadItemId!, props) : undefined,
        {
          target_name: !!targetFilename ? targetFilename : itemInfo.name,
          target_directory: targetPath,
          priority: API.QueuePriorityEnum.DEFAULT,
        },
        session,
      );
    } catch (e) {
      NotificationActions.error({
        title: t(toI18nKey('queueingFailed', UI.Modules.COMMON), {
          defaultValue: 'Failed to queue the item {{item.name}}',
          replace: {
            item: itemInfo,
          },
        }),
        message: e.message,
      });
    }

    runBackgroundSocketAction(
      () => addHistory(HistoryStringEnum.DOWNLOAD_DIR, targetPath),
      t,
    );

    handleClose();
  };

  const getInitialBrowsePath = () => {
    const path = historyPaths.length > 0 ? historyPaths[historyPaths.length - 1] : '';
    return itemInfo.type.id === 'directory' ? path : path + itemInfo.name;
  };

  const handleBrowse = () => {
    navigate(`browse`);
  };

  const commonDialogProps: ModalProps = {
    subHeader: itemInfo.name,
    title: translate('Download', t, UI.Modules.COMMON),
    icon: IconConstants.DOWNLOAD,
    closable: false,
  };

  const hasFileBrowserAccess = LoginStore.hasAccess(API.AccessEnum.FILESYSTEM_VIEW);
  return (
    <Routes>
      <Route
        path={`browse`}
        element={
          <FileBrowserDialog
            onConfirm={(path, directoryPath, fileName) =>
              handleDownload(directoryPath, fileName)
            }
            initialPath={getInitialBrowsePath()}
            selectMode={
              itemInfo.type.id === 'directory'
                ? UI.FileSelectModeEnum.DIRECTORY
                : UI.FileSelectModeEnum.FILE
            }
            historyId={FilesystemConstants.LOCATION_DOWNLOAD}
            approveCaption={translate('Download', t, UI.Modules.COMMON)}
            {...commonDialogProps}
          />
        }
      />
      <Route
        index
        element={
          <Modal
            className="download-dialog"
            fullHeight={true}
            {...commonDialogProps}
            {...other}
          >
            <DownloadLayout
              downloadHandler={handleDownload}
              handleBrowse={hasFileBrowserAccess ? handleBrowse : undefined}
              historyPaths={historyPaths}
              favoritePaths={favoritePaths}
              sharePaths={sharePaths}
              itemInfo={itemInfo}
            />
          </Modal>
        }
      />
    </Routes>
  );
};

export default ModalRouteDecorator<DownloadDialogProps>(
  DataProviderDecorator<
    DownloadDialogProps & DownloadDialogRouteProps,
    DownloadDialogDataProps
  >(DownloadDialog, {
    urls: {
      sharePaths: ShareConstants.GROUPED_ROOTS_GET_URL,
      favoritePaths: FavoriteDirectoryConstants.GROUPED_DIRECTORIES_URL,
      historyPaths: HistoryConstants.STRINGS_URL + '/' + HistoryStringEnum.DOWNLOAD_DIR,
      itemInfo: ({ params, itemDataGetter }, socket) => {
        return itemDataGetter(params.downloadItemId!, socket);
      },
    },
  }),
  '/download/:downloadItemId',
);
