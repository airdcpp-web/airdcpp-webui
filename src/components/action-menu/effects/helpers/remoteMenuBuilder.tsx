import * as API from '@/types/api';
import * as UI from '@/types/ui';

import MenuConstants from '@/constants/MenuConstants';
import IconConstants from '@/constants/IconConstants';
import { RemoteMenu, RemoteMenuData } from './remoteMenuFetcher';
import { UILogger } from '@/effects/LoggerEffect';

export interface CombinedContextMenuItem {
  typeData: RemoteMenuData;
  item: API.ContextMenuItem;
}

export interface CombinedContextMenu extends Omit<API.GroupedContextMenu, 'items'> {
  items: CombinedContextMenuItem[];
}

interface MenuOptions {
  handleSelect: SelectHandler;
  nestingThreshold: number;
}

type SelectHandler = (
  menuItem: API.ContextMenuItem,
  remoteMenuData: RemoteMenuData,
) => void;

const toMenuItem = (
  menuItem: API.ContextMenuItem,
  remoteMenuData: RemoteMenuData,
  handleSelect: SelectHandler,
): UI.ActionMenuItem => {
  const onClick = () => {
    handleSelect(menuItem, remoteMenuData);
  };

  const { id, icon, title } = menuItem;
  return {
    id,
    item: {
      onClick,
      icon: icon[MenuConstants.ICON_TYPE_ID],
      children: title,
    },
  };
};

export const remoteMenuToActionMenuItems = (
  menu: CombinedContextMenu,
  { handleSelect, nestingThreshold }: MenuOptions,
): UI.ActionMenuItem[] => {
  const items = menu.items.map(({ item, typeData }) =>
    toMenuItem(item, typeData, handleSelect),
  );

  if (items.length <= nestingThreshold) {
    return items;
  }

  return [
    {
      id: menu.id,
      item: {
        icon: menu.icon[MenuConstants.ICON_TYPE_ID] || IconConstants.EXTENSION,
        onClick: () => {
          // ..
        },
        children: <>{menu.title}</>,
      },
      children: items,
    },
  ];
};

// Combine menus by their subscriber ID and remove duplicate items
export const createRemoteMenuCombineReducer = (logger: UILogger) => {
  return (reduced: CombinedContextMenu[], typeMenus: RemoteMenu | undefined) => {
    if (!typeMenus) {
      return reduced;
    }

    for (const menu of typeMenus.menus) {
      const existing = reduced.find((m) => m.id === menu.id);

      const toMenuItem = (item: API.ContextMenuItem): CombinedContextMenuItem => ({
        typeData: typeMenus.typeData,
        item,
      });

      if (existing) {
        existing.items.push(
          ...menu.items
            .filter((item) => {
              const existingItem = existing.items.find(
                (existingItem) =>
                  existingItem.item.id === item.id &&
                  existingItem.item.hook_id === item.hook_id,
              );
              if (existingItem) {
                logger.info(
                  // eslint-disable-next-line max-len
                  `Removing a duplicate item ${item.id} (subscriber ${menu.id}) from menu ${typeMenus.typeData.remoteMenuId} (exists for menu ${existingItem.typeData.remoteMenuId})`,
                );
              }

              return !existingItem;
            })
            .map(toMenuItem),
        );
      } else {
        reduced.push({
          ...menu,
          items: menu.items.map(toMenuItem),
        });
      }
    }

    return reduced;
  };
};
