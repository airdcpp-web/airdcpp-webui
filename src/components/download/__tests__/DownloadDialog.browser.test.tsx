import { afterEach, beforeEach, describe, expect, test, vi } from 'vitest';
import { waitFor, fireEvent } from '@testing-library/react';

import { getMockServer } from 'airdcpp-apisocket/tests';

import ShareConstants from '@/constants/ShareConstants';
import { ShareGetGroupedRootsResponse } from '@/tests/mocks/api/share';
import FavoriteDirectoryConstants from '@/constants/FavoriteDirectoryConstants';
import { FavoriteDirectoriesGroupedPathsResponse } from '@/tests/mocks/api/favorite-directories';
import HistoryConstants, { HistoryStringEnum } from '@/constants/HistoryConstants';
import {
  FilelistGetFilelistItemFileResponse,
  FilelistGetResponse,
  MOCK_FILELIST_ITEM_ID,
} from '@/tests/mocks/api/filelist';
import FilesystemConstants from '@/constants/FilesystemConstants';
import {
  FilesystemDiskInfoResponse,
  FilesystemListContentResponse,
} from '@/tests/mocks/api/filesystem';
import { renderDataRoutes } from '@/tests/render/test-renderers';
import DownloadDialog from '../DownloadDialog';

import * as UI from '@/types/ui';
import * as API from '@/types/api';

import { MockHintedUser1Response } from '@/tests/mocks/api/user';

import { HistoryStringPathResponse } from '@/tests/mocks/api/history';
import {
  createTestRouteModalController,
  TestRouteModalNavigateButton,
} from '@/tests/helpers/test-dialog-helpers';
import { waitForData } from '@/tests/helpers/test-helpers';
import { initCommonDataMocks } from '@/tests/mocks/mock-data-common';

// tslint:disable:no-empty
describe('DownloadDialog', () => {
  let server: ReturnType<typeof getMockServer>;
  const getSocket = async () => {
    const commonData = await initCommonDataMocks(server);

    // Target paths fetch
    server.addRequestHandler(
      'GET',
      ShareConstants.GROUPED_ROOTS_GET_URL,
      ShareGetGroupedRootsResponse,
    );
    server.addRequestHandler(
      'GET',
      FavoriteDirectoryConstants.GROUPED_DIRECTORIES_URL,
      FavoriteDirectoriesGroupedPathsResponse,
    );
    server.addRequestHandler(
      'GET',
      `${HistoryConstants.STRINGS_URL}/${HistoryStringEnum.DOWNLOAD_DIR}`,
      HistoryStringPathResponse,
    );

    // Saving of the selected path
    server.addRequestHandler(
      'POST',
      `${HistoryConstants.STRINGS_URL}/${HistoryStringEnum.DOWNLOAD_DIR}`,
      undefined,
    );

    // File browser
    server.addRequestHandler(
      'POST',
      FilesystemConstants.DISK_INFO_URL,
      FilesystemDiskInfoResponse,
    );
    server.addRequestHandler(
      'POST',
      FilesystemConstants.LIST_URL,
      FilesystemListContentResponse,
    );

    return commonData;
  };

  const renderDialog = async () => {
    const commonData = await getSocket();
    const handleDownload = vi.fn<UI.DownloadHandler<API.FilelistItem>>();

    const DownloadDialogTest = () => {
      return (
        <>
          <TestRouteModalNavigateButton
            modalRoute={`/home/download/${MOCK_FILELIST_ITEM_ID}`}
          />
          <DownloadDialog
            downloadHandler={handleDownload}
            itemDataGetter={() =>
              Promise.resolve(FilelistGetFilelistItemFileResponse as API.FilelistItem)
            }
            userGetter={() => MockHintedUser1Response as API.HintedUser}
            sessionItem={FilelistGetResponse}
          />
        </>
      );
    };

    const routes = [
      {
        path: '/home/*',
        Component: DownloadDialogTest,
      },
    ];

    const renderData = renderDataRoutes(routes, commonData, {
      routerProps: { initialEntries: ['/home'] },
    });

    const modalController = createTestRouteModalController(renderData);
    return { ...commonData, handleDownload, modalController, ...renderData };
  };

  const expectDownloadHandlerToMatchSnapshot = (
    handleDownload: ReturnType<typeof vi.fn<UI.DownloadHandler<API.FilelistItem>>>,
  ) => {
    expect(handleDownload).toHaveBeenCalledTimes(1);
    expect(handleDownload.mock.calls[0][0]).toMatchSnapshot();
    expect(handleDownload.mock.calls[0][1]).toMatchSnapshot();
  };

  beforeEach(() => {
    server = getMockServer();
  });

  afterEach(() => {
    server.stop();
    localStorage.clear();
  });

  test('should load and close', async () => {
    const { getByText, socket, modalController } = await renderDialog();

    await modalController.openDialog();

    // Check content
    await waitFor(() => expect(getByText('Download')).toBeTruthy());

    await modalController.closeDialogButton('Close');

    socket.disconnect();
  });

  test('should handle download', async () => {
    const { socket, modalController, handleDownload } = await renderDialog();

    await modalController.openDialog();

    // Download
    await modalController.closeDialogText(HistoryStringPathResponse[0]);

    // Check
    expectDownloadHandlerToMatchSnapshot(handleDownload);

    socket.disconnect();
  });

  test('should open and close download dialog', async () => {
    const { socket, modalController, getByText, queryByText } = await renderDialog();

    // Open download file browser
    await modalController.openDialog();
    expect(fireEvent.click(getByText('Browse'))).toBeTruthy();

    // Open browse dialog
    await waitForData(/Loading items/i, queryByText);

    // Close it
    await modalController.closeDialogButton('Cancel');

    socket.disconnect();
  });

  test('should handle browse dialog download', async () => {
    const { socket, modalController, handleDownload, getByText, queryByText } =
      await renderDialog();

    // Open download file browser
    await modalController.openDialog();
    expect(fireEvent.click(getByText('Browse'))).toBeTruthy();

    // Open browse dialog
    await waitForData('Loading items', queryByText);

    // Download
    await modalController.closeDialogButton('Download');

    expectDownloadHandlerToMatchSnapshot(handleDownload);

    socket.disconnect();
  });
});
