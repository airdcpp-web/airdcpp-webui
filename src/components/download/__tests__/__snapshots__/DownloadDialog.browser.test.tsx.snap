// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DownloadDialog > should handle browse dialog download 1`] = `
{
  "entity": {
    "id": "HPI67YG335U72FUS626M6WYFGR5EBZTNCEJU3JA",
    "location": null,
    "partial_list": true,
    "read": false,
    "share_profile": null,
    "state": {
      "id": "download_pending",
      "str": "Download pending",
      "time_finished": 0,
    },
    "total_files": 936066,
    "total_size": 133137630299481,
    "user": {
      "cid": "HNLUODI2YZ2U7FDWMHFNJU65ERGKUN4MH7GW5LY",
      "flags": [
        "asch",
        "ccpm",
      ],
      "hub_names": "ADC hub 1 (ADC hub 3)",
      "hub_url": "adcs://hub1.adc/?kp=SHA256/H4DUKGI3U3LQHMZXU6SLQA7LF5RVD6HJFGYZBN2R5RRQ5A2RZ4ZA",
      "hub_urls": [
        "adcs://hub1.adc/?kp=SHA256/H4DUKGI3U3LQHMZXU6SLQA7LF5RVD6HJFGYZBN2R5RRQ5A2RZ4ZA",
        "adcs://hub3.adc:2482",
      ],
      "nicks": "[Hub1]User1 ([Hub3]User1)",
    },
  },
  "itemInfo": {
    "complete": true,
    "dupe": null,
    "id": 323,
    "name": "ubuntu-20.04.2.0-desktop-amd64.iso",
    "path": "/Downloads/ubuntu-20.04.2.0-desktop-amd64.iso",
    "size": 2877227008,
    "time": 1618062636,
    "tth": "2E2P2K5JP6KARHNFGUMIFENN4SGBENKLPGQNUAA",
    "type": {
      "content_type": "",
      "id": "file",
      "str": "iso",
    },
  },
  "user": {
    "cid": "HNLUODI2YZ2U7FDWMHFNJU65ERGKUN4MH7GW5LY",
    "flags": [
      "asch",
      "ccpm",
    ],
    "hub_names": "ADC hub 1 (ADC hub 3)",
    "hub_url": "adcs://hub1.adc/?kp=SHA256/H4DUKGI3U3LQHMZXU6SLQA7LF5RVD6HJFGYZBN2R5RRQ5A2RZ4ZA",
    "hub_urls": [
      "adcs://hub1.adc/?kp=SHA256/H4DUKGI3U3LQHMZXU6SLQA7LF5RVD6HJFGYZBN2R5RRQ5A2RZ4ZA",
      "adcs://hub3.adc:2482",
    ],
    "nicks": "[Hub1]User1 ([Hub3]User1)",
  },
}
`;

exports[`DownloadDialog > should handle browse dialog download 2`] = `
{
  "priority": -1,
  "target_directory": "/mnt/disk1/Images/",
  "target_name": "ubuntu-20.04.2.0-desktop-amd64.iso",
}
`;

exports[`DownloadDialog > should handle download 1`] = `
{
  "entity": {
    "id": "HPI67YG335U72FUS626M6WYFGR5EBZTNCEJU3JA",
    "location": null,
    "partial_list": true,
    "read": false,
    "share_profile": null,
    "state": {
      "id": "download_pending",
      "str": "Download pending",
      "time_finished": 0,
    },
    "total_files": 936066,
    "total_size": 133137630299481,
    "user": {
      "cid": "HNLUODI2YZ2U7FDWMHFNJU65ERGKUN4MH7GW5LY",
      "flags": [
        "asch",
        "ccpm",
      ],
      "hub_names": "ADC hub 1 (ADC hub 3)",
      "hub_url": "adcs://hub1.adc/?kp=SHA256/H4DUKGI3U3LQHMZXU6SLQA7LF5RVD6HJFGYZBN2R5RRQ5A2RZ4ZA",
      "hub_urls": [
        "adcs://hub1.adc/?kp=SHA256/H4DUKGI3U3LQHMZXU6SLQA7LF5RVD6HJFGYZBN2R5RRQ5A2RZ4ZA",
        "adcs://hub3.adc:2482",
      ],
      "nicks": "[Hub1]User1 ([Hub3]User1)",
    },
  },
  "itemInfo": {
    "complete": true,
    "dupe": null,
    "id": 323,
    "name": "ubuntu-20.04.2.0-desktop-amd64.iso",
    "path": "/Downloads/ubuntu-20.04.2.0-desktop-amd64.iso",
    "size": 2877227008,
    "time": 1618062636,
    "tth": "2E2P2K5JP6KARHNFGUMIFENN4SGBENKLPGQNUAA",
    "type": {
      "content_type": "",
      "id": "file",
      "str": "iso",
    },
  },
  "user": {
    "cid": "HNLUODI2YZ2U7FDWMHFNJU65ERGKUN4MH7GW5LY",
    "flags": [
      "asch",
      "ccpm",
    ],
    "hub_names": "ADC hub 1 (ADC hub 3)",
    "hub_url": "adcs://hub1.adc/?kp=SHA256/H4DUKGI3U3LQHMZXU6SLQA7LF5RVD6HJFGYZBN2R5RRQ5A2RZ4ZA",
    "hub_urls": [
      "adcs://hub1.adc/?kp=SHA256/H4DUKGI3U3LQHMZXU6SLQA7LF5RVD6HJFGYZBN2R5RRQ5A2RZ4ZA",
      "adcs://hub3.adc:2482",
    ],
    "nicks": "[Hub1]User1 ([Hub3]User1)",
  },
}
`;

exports[`DownloadDialog > should handle download 2`] = `
{
  "priority": -1,
  "target_directory": "/home/airdcpp/Downloads/",
  "target_name": "ubuntu-20.04.2.0-desktop-amd64.iso",
}
`;
